retrofit/src/main/java/retrofit/MethodInfo.java

```diff
@@ -223,69 +210,35 @@ private void parseHttpMethodAndPath(String method, String path, boolean hasBody)
     return builder.build();
   }
 
-  /** Loads {@link #responseObjectType}. */
-  private ExecutionType parseResponseType() {
-    // Synchronous methods have a non-void return type.
-    // Observable methods have a return type of Observable.
+  /** Loads {@link #adapter}. */
+  private void parseResponseType() {
     Type returnType = method.getGenericReturnType();
 
-    // Asynchronous methods should have a Callback type as the last argument.
-    Type lastArgType = null;
-    Class<?> lastArgClass = null;
-    Type[] parameterTypes = method.getGenericParameterTypes();
-    if (parameterTypes.length > 0) {
-      Type typeToCheck = parameterTypes[parameterTypes.length - 1];
-      lastArgType = typeToCheck;
-      if (typeToCheck instanceof ParameterizedType) {
-        typeToCheck = ((ParameterizedType) typeToCheck).getRawType();
-      }
-      if (typeToCheck instanceof Class) {
-        lastArgClass = (Class<?>) typeToCheck;
-      }
-    }
-
-    boolean hasReturnType = returnType != void.class;
-    boolean hasCallback = lastArgClass != null && Callback.class.isAssignableFrom(lastArgClass);
-
     // Check for invalid configurations.
-    if (hasReturnType && hasCallback) {
-      throw methodError("Must have return type or Callback as last argument, not both.");
-    }
-    if (!hasReturnType && !hasCallback) {
-      throw methodError("Must have either a return type or Callback as last argument.");
+    if (returnType == void.class) {
+      throw methodError("Service methods cannot return void.");
     }
 
-    if (hasReturnType) {
-      if (Platform.HAS_RX_JAVA) {
-        Class rawReturnType = Types.getRawType(returnType);
-        if (RxSupport.isObservable(rawReturnType)) {
-          returnType = RxSupport.getObservableType(returnType, rawReturnType);
-          responseObjectType = getParameterUpperBound((ParameterizedType) returnType);
-          return ExecutionType.RX;
-        }
+    //noinspection ForLoopReplaceableByForEach
+    for (int i = 0, count = factories.size(); i < count; i++) {
+      CallAdapter.Factory factory = factories.get(i);
+      CallAdapter adapter = factory.get(returnType);
+      if (adapter != null) {
+        this.adapter = adapter;
+        return;
       }
-      responseObjectType = returnType;
-      return ExecutionType.SYNC;
     }
 
-    lastArgType = Types.getSupertype(lastArgType, Types.getRawType(lastArgType), Callback.class);
-    if (lastArgType instanceof ParameterizedType) {
-      responseObjectType = getParameterUpperBound((ParameterizedType) lastArgType);
-      return ExecutionType.ASYNC;
+    StringBuilder builder = new StringBuilder() //
+        .append("No registered call adapters were able to handle return type ")
+        .append(returnType)
+        .append(". Checked: [ ");
+    for (int i = 0, count = factories.size(); i < count; i++) {
```

Comment:
You might be better off just using list's toString. If for whatever reason I've got my own configurable factory class, this prevents me from adding my own toString debugging. (Especially since you could implement toString() to return the factory class simplename in the built-in implementations.)